### PR TITLE
Annotation Feature: Positive and Negative Annotation. 

### PR DESCRIPTION
**Pull request by imminal35**
This PR contains Annotation feature for Vador framework. Currently this will add 2 annotations. 
1. Positive 
2. Negative 

A user can use these on field level and needs to pass 2 params while consuming. 
1. Annotation (Positive or Negative)
2. failureKey (String) 

Also, As part of these I have also added 3 tests which covers all possible flow. 

Below snapshots shows successful build after new changes. 
Screenshot 2023-09-01 at 6 15 38 PM

### CODE REVIEW

File: vador/src/main/java/com/salesforce/vador/execution/strategies/FailFastStrategies.kt [hunk 1]
```diff
@@ -90,7 +97,7 @@ internal fun <FailureT, ValidatableT> failFastForEach(
   findAndFilterInvalids(
       validatables,
       failureForNullValidatable,
-      batchValidationConfig.findAndFilterDuplicatesConfigs
+      batchValidationConfig.findAndFilterDuplicatesConfigs,
```
Comment (overfullstack) [inline vador/src/main/java/com/salesforce/vador/execution/strategies/FailFastStrategies.kt:100]: Pls Remove this formatting changes

File: vador/src/main/java/com/salesforce/vador/types/Validators.kt [hunk 2]
```diff
@@ -14,3 +14,7 @@ fun interface Validator<ValidatableT, FailureT> : CheckedFunction1<ValidatableT,
 
 fun interface ValidatorEtr<ValidatableT, FailureT> :
   CheckedFunction1<Either<FailureT?, ValidatableT?>, Either<FailureT?, *>>
+
+fun interface ValidatorAnnotation<FailureT> {
```
Comment (overfullstack) [inline vador/src/main/java/com/salesforce/vador/types/Validators.kt:18]: Rename this to `ValidatorAnnotation1`

File: vador/src/main/java/com/salesforce/vador/execution/strategies/FailFastStrategies.kt (hunk 1 above)
Comment (overfullstack) [reply vador/src/main/java/com/salesforce/vador/execution/strategies/FailFastStrategies.kt:100]: I ran spotlessApply locally and it didnt append these commas, please revert them and try that command again. Let me know if you see them agian

Comment (imminal35) [reply vador/src/main/java/com/salesforce/vador/execution/strategies/FailFastStrategies.kt]: Done

File: vador/src/main/java/com/salesforce/vador/types/Validators.kt (hunk 2 above)
Comment (imminal35) [reply vador/src/main/java/com/salesforce/vador/types/Validators.kt]: Done

File: vador/src/main/java/com/salesforce/vador/execution/strategies/AnnotationProcessorBase.kt [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        var validator: Validator<ValidatableT?, FailureT?>? = null
+        val map = mapOfAnnotation?._1()
+        val none = mapOfAnnotation?._2()
+        if (field.isAnnotationPresent(ValidateWith::class.java)) {
+          val annotation =
+            field.getAnnotation(
+              ValidateWith::class.java,
+            )
+          val failureKey = annotation.failureKey
+          if (ValidatorAnnotation::class.java.isAssignableFrom(annotation.validator.java)) {
```
[the remark this answers is not in the document]
Comment (imminal35) [reply vador/src/main/java/com/salesforce/vador/execution/strategies/AnnotationProcessorBase.kt]: Done

File: vador/src/main/java/com/salesforce/vador/execution/strategies/AnnotationProcessorBase.kt [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    NoSuchMethodException::class,
+    InvocationTargetException::class,
+    InstantiationException::class,
+    IllegalAccessException::class,
+  )
+  fun <ValidatableT, FailureT> derivedValidators(
+    bean: Any?,
+    mapOfAnnotation: Tuple2<MutableMap<String, FailureT>, FailureT>?
+  ): List<ValidatorEtr<ValidatableT?, FailureT?>> {
+    val listOfValidatorEtr: MutableList<ValidatorEtr<ValidatableT?, FailureT?>> = ArrayList()
```
[the remark this answers is not in the document]
Comment (imminal35) [reply vador/src/main/java/com/salesforce/vador/execution/strategies/AnnotationProcessorBase.kt]: Done

File: vador/src/main/java/com/salesforce/vador/execution/strategies/AnnotationProcessorBase.kt [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+object AnnotationProcessorBase {
+  @Throws(
+    NoSuchMethodException::class,
+    InvocationTargetException::class,
+    InstantiationException::class,
+    IllegalAccessException::class,
+  )
+  fun <ValidatableT, FailureT> derivedValidators(
+    bean: Any?,
```
[the remark this answers is not in the document]
Comment (imminal35) [reply vador/src/main/java/com/salesforce/vador/execution/strategies/AnnotationProcessorBase.kt]: Done

File: vador/src/test/java/com/salesforce/vador/execution/VadorAnnotationTest.java
[the remark this answers is not in the document]
Comment (imminal35) [reply vador/src/test/java/com/salesforce/vador/execution/VadorAnnotationTest.java:1]: Done

File: vador/src/main/java/com/salesforce/vador/config/base/BaseValidationConfig.java [hunk 6]
```diff
@@ -75,6 +75,8 @@ public abstract class BaseValidationConfig<ValidatableT, FailureT> {
           @NonNull FailureT>
       withValidators;
 
+  @Nullable Tuple2<Map<String, @Nullable FailureT>, @Nullable FailureT> forAnnotation;
```
[the remark this answers is not in the document]
Comment (imminal35) [reply vador/src/main/java/com/salesforce/vador/config/base/BaseValidationConfig.java]: Done

File: vador/src/main/java/com/salesforce/vador/annotation/Positive.kt [hunk 7]
```diff
@@ -0,0 +1,17 @@
+package com.salesforce.vador.annotation
+
+import com.salesforce.vador.types.ValidatorAnnotation
+
+class Positive<FailureT> : ValidatorAnnotation<FailureT> {
+  override fun validate(value: Any, failure: FailureT, none: FailureT): FailureT {
+    val intValue =
+      when (value) {
+        is Int -> value
+        else -> throw IllegalArgumentException("Value is not an integer")
```
[the remark this answers is not in the document]
Comment (imminal35) [reply vador/src/main/java/com/salesforce/vador/annotation/Positive.kt]: Done

File: vador/src/main/java/com/salesforce/vador/annotation/NonNegative.kt [hunk 8]
```diff
@@ -0,0 +1,17 @@
+package com.salesforce.vador.annotation
+
+import com.salesforce.vador.types.ValidatorAnnotation
+
+class NonNegative<FailureT> : ValidatorAnnotation<FailureT> {
+  override fun validate(value: Any, failure: FailureT, none: FailureT): FailureT {
+    val intValue =
+      when (value) {
+        is Int -> value
+        else -> throw IllegalArgumentException("Value is not an integer")
```
[the remark this answers is not in the document]
Comment (imminal35) [reply vador/src/main/java/com/salesforce/vador/annotation/NonNegative.kt]: Done

File: vador/src/main/java/com/salesforce/vador/annotation/Negative.kt [hunk 9]
```diff
@@ -0,0 +1,17 @@
+package com.salesforce.vador.annotation
+
+import com.salesforce.vador.types.ValidatorAnnotation
+
+class Negative<FailureT> : ValidatorAnnotation<FailureT> {
+  override fun validate(value: Any, failure: FailureT, none: FailureT): FailureT {
+    val intValue =
+      when (value) {
+        is Int -> value
+        else -> throw IllegalArgumentException("Value is not an integer")
```
[the remark this answers is not in the document]
Comment (imminal35) [reply vador/src/main/java/com/salesforce/vador/annotation/Negative.kt]: Done

File: vador/src/main/java/com/salesforce/vador/execution/strategies/AnnotationProcessorBase.kt [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+              validatorAnnotation.validate(
+                field,
+                FieldUtils.readField(field, bean, true) as Any,
+                map?.get(failureKey),
+                none,
+              )
+            }
+          liftToEtr<FailureT, ValidatableT>(validator, none)
+        } else {
+          throw IllegalArgumentException("Provided Annotation is not supported.")
```
Comment (overfullstack) [inline vador/src/main/java/com/salesforce/vador/execution/strategies/AnnotationProcessorBase.kt:45]: Please add more details to the message to make it more clear about the corrective action consumer needs to take

File: vador/src/main/java/com/salesforce/vador/execution/strategies/AnnotationProcessorBase.kt [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+import com.salesforce.vador.annotation.ValidateWith
+import com.salesforce.vador.lift.liftToEtr
+import com.salesforce.vador.types.Validator
+import com.salesforce.vador.types.ValidatorAnnotation1
+import com.salesforce.vador.types.ValidatorEtr
+import io.vavr.Tuple2
+import java.lang.reflect.InvocationTargetException
+import org.apache.commons.lang3.reflect.FieldUtils
+
+object AnnotationProcessorBase {
```
Comment (overfullstack) [inline vador/src/main/java/com/salesforce/vador/execution/strategies/AnnotationProcessorBase.kt:12]: `AnnotationProcessorBase` - Pls remove `Base` from the name as this is not ancestor for any class

File: vador/src/test/java/com/salesforce/vador/execution/VadorAnnotationTest.java [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+  private static final BeanCustom2 VALIDATBLECUSTOM2 =
+      new BeanCustom2(new ID("Test Class1"), new ID("Test Class2"));
+
+  private static final BeanCustom3 VALIDATBLECUSTOM3 = new BeanCustom3(new ID("Test Class1"));
+
+  @Test
+  void failFastWithFirstFailureWithValidatorAnnotation() {
+    final var validationConfig =
+        ValidationConfig.<VadorAnnotationTest.Bean, ValidationFailure>toValidate()
+            .forAnnotations(Tuple.of(Map.of("Unexpected_exception", UNKNOWN_EXCEPTION), NONE))
```
Comment (overfullstack) [inline vador/src/test/java/com/salesforce/vador/execution/VadorAnnotationTest.java:40]: NIT: Follow cameCase, like `unexpectedException`

**Review by overfullstack — APPROVED** [review]

**Pull request merged**